plugins/home/src/components/CustomHomepage/CustomHomepageGrid.tsx

```diff
@@ -0,0 +1,345 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { ReactNode, useCallback, useMemo } from 'react';
+import { Layout, Layouts, Responsive, WidthProvider } from 'react-grid-layout';
+import {
+  storageApiRef,
+  useApi,
+  getAllComponentData,
+  getComponentData,
+} from '@backstage/core-plugin-api';
+import 'react-grid-layout/css/styles.css';
+import 'react-resizable/css/styles.css';
+import { createStyles, Dialog, makeStyles, Theme } from '@material-ui/core';
+import { compact } from 'lodash';
+import useObservable from 'react-use/lib/useObservable';
+import { ContentHeader, ErrorBoundary } from '@backstage/core-components';
+import Typography from '@material-ui/core/Typography';
+import { WidgetSettings } from '../../extensions';
+import { WidgetSettingsOverlay } from './WidgetSettingsOverlay';
+import { AddWidgetDialog } from './AddWidgetDialog';
+import { CustomHomepageButtons } from './CustomHomepageButtons';
+
+// eslint-disable-next-line new-cap
+const ResponsiveGrid = WidthProvider(Responsive);
+const hash = require('object-hash');
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    responsiveGrid: {
+      '& .react-grid-item > .react-resizable-handle': {
+        mixBlendMode: 'difference',
+        borderRight: '2px solid rgba(255, 255, 255, 0.7)',
+        borderBottom: '2px solid rgba(255, 255, 255, 0.5)',
+      },
+    },
+    contentHeaderBtn: {
+      marginLeft: theme.spacing(2),
+    },
+    widgetWrapper: {
+      '& > *:first-child': {
+        width: '100%',
+        height: '100%',
+      },
+      '& + .react-grid-placeholder': {
+        backgroundColor: theme.palette.primary.light,
+      },
+      '&.edit > :active': {
+        cursor: 'move',
+      },
+    },
+  }),
+);
+
+export type Widget = WidgetSettings & {
+  name: string;
+  title?: string;
+  component: ReactNode & {
+    type: any;
+    props: any;
+  };
+};
+
+type WidgetStoreItem = {
+  id: string;
+  layout: Layout;
+  settings: Record<string, any>;
+};
+
+const getAvailableWidgets = (children: ReactNode[]): Widget[] => {
+  const ret = children.map(child => {
+    const componentData = getAllComponentData(child);
+    if (!componentData) {
+      return null;
+    }
+    return {
+      component: child,
+      ...Object.fromEntries(componentData as Iterable<any>),
+    } as Widget;
+  });
+  return compact(ret);
+};
+
+function useHomeStorage(
+  defaultWidgets: WidgetStoreItem[],
+): [WidgetStoreItem[], (value: WidgetStoreItem[]) => void] {
+  const storageApi = useApi(storageApiRef);
+  const homeKey = 'home';
+  // TODO: Support multiple home pages
+  const setWidgets = useCallback(
+    (value: WidgetStoreItem[]) => {
```

Comment:
I think it's worth to think upfront about versioning here. Any model that is persisted anywhere, will benefit from having that in mind from the get go.

Some considerations:

- Store the data in a bucket with a specific, descriptive name. It could be something like `home.customHomepageGrid` at least
- Have a types.ts file that houses specific storage model info types that can be kept static over time all collected in one place
- Make a versioned root object type, like

  ```ts
  export interface CustomHomepageGridStateV1 {
    version: 1;
    // even if there's only a "default" key for now
    pages: Record<string, {
      widgets: Array<....>;
    }
  }
  ```
  
  or something like that, which is what's being serialized/deserialized as JSON. Now in a future code change you may introduce a V2 variant of that, but the loader code still has the ability to load the V1 and convert on-the-fly to V2 which is what the view code had been updated to understand.
- Stay strict about keeping those models intact, don't mix too much with other code; it's your de facto storage model and what gets persisted may stay forever
- You may even consider defining these as zod schemas, and inferring the typesript type from that. Now you can do strict runtime validation of what goes in and out of storage, which is a very nice stability guarantee. [See here](https://github.com/backstage/backstage/blob/5b975591efecc2a91ed5aae4965630aecf8a9634/packages/backend-tasks/src/tasks/types.ts#L390) for an example of that being done in the tasks system.

Reply:
Good idea. The root object is a good way to handle possible future updates that might have breaking changes to the way the configuration is saved. I will dig into this as well. Thanks for good points!

Reply:
Thanks for the pointers, should be much safer now, please check!